File: src/lib/session-resume.ts
import { detectHostPlatform, type HostPlatform } from '@/lib/platform-detect';

const MAX_SESSION_ID_LEN = 512;

/** Planned native resume argv (`argv[0]` is the CLI name). */
export type NativeResumePlan = {
  agentId: string;
  argv: string[];
};

/**
 * Official **interactive TUI** resume argv for a listed native session id.
 * Flags match herdr `agent_resume.rs` (`claude --resume`, `codex resume`, …).
 * Chat subsequent turns use print-mode resume in core, not this helper.
 */
export function planNativeResume(
  agentId: string,
  sessionId: string | null | undefined,
  host: HostPlatform = detectHostPlatform(),
): NativeResumePlan | null {
  const id = validSessionId(sessionId);
  if (!id) return null;
  const argv = resumeArgv(agentId, id, host);
  if (!argv) return null;
  return { agentId, argv };
}

export function formatResumeCommand(argv: readonly string[]): string {
  return argv.map(quoteArg).join(' ');
}

export function nativeResumeCommand(
  agentId: string,
  sessionId: string | null | undefined,
  host?: HostPlatform,
): string | null {
  const plan = planNativeResume(agentId, sessionId, host);
  return plan ? formatResumeCommand(plan.argv) : null;
}

function resumeArgv(agentId: string, sessionId: string, host: HostPlatform): string[] | null {
  switch (agentId) {
    case 'claude':
      return ['claude', '--resume', sessionId];
    case 'codex':
      return ['codex', 'resume', sessionId];
    case 'kimi':
      return ['kimi', '--session', sessionId];
    case 'grok':
      return ['grok', '--resume', sessionId];
    case 'pi':
      return ['pi', '--session', sessionId];
    case 'cursor':
      return [host === 'windows' ? 'cursor-agent.cmd' : 'cursor-agent', '--resume', sessionId];
    default:
      return null;
  }
}

function validSessionId(value: string | null | undefined): string | null {
  const id = value?.trim() ?? '';
  if (!id || id.length > MAX_SESSION_ID_LEN || /[\u0000-\u001f\u007f]/.test(id)) {
    return null;
  }
  return id;
}

function quoteArg(value: string): string {
  if (/^[A-Za-z0-9._:/=+-]+$/.test(value)) return value;
  return `"${value.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`;
}

File: crates/agenthub-core/src/adapters/session_resume.rs
//! Official CLI argv for resuming a native agent session.
//!
//! Interactive flags match herdr's `agent_resume.rs` (`claude --resume`,
//! `codex resume`, …). Chat subsequent turns use print-mode resume
//! (`claude -p --resume`, `codex exec resume`) via [`supports_print_resume`].

use crate::models::AgentId;

const MAX_SESSION_ID_LEN: usize = 512;

/// Planned native resume invocation (`program` is `argv[0]`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NativeResumePlan {
    pub agent: AgentId,
    pub argv: Vec<String>,
}

/// Build the official **interactive TUI** resume argv for a listed native session id.
///
/// Returns `None` when the id is unusable or the agent has no known resume CLI.
pub fn plan_native_resume(agent: AgentId, session_id: &str) -> Option<NativeResumePlan> {
    let session_id = valid_session_id(session_id)?;
    let argv = resume_argv(agent, session_id)?;
    Some(NativeResumePlan { agent, argv })
}

/// Whether Chat can continue this agent with print-mode resume (not TUI).
pub fn supports_print_resume(agent: AgentId) -> bool {
    matches!(agent, AgentId::Claude | AgentId::Codex)
}

fn resume_argv(agent: AgentId, session_id: &str) -> Option<Vec<String>> {
    let argv = match agent {
        AgentId::Claude => vec!["claude".into(), "--resume".into(), session_id.into()],
        AgentId::Codex => vec!["codex".into(), "resume".into(), session_id.into()],
        AgentId::Kimi => vec!["kimi".into(), "--session".into(), session_id.into()],
        AgentId::Grok => vec!["grok".into(), "--resume".into(), session_id.into()],
        AgentId::Pi => vec!["pi".into(), "--session".into(), session_id.into()],
        AgentId::Cursor => vec![
            cursor_program().into(),
            "--resume".into(),
            session_id.into(),
        ],
        AgentId::WorkBuddy | AgentId::Dsh => return None,
    };
    Some(argv)
}

fn cursor_program() -> &'static str {
    if cfg!(windows) {
        "cursor-agent.cmd"
    } else {
        "cursor-agent"
    }
}

pub(crate) fn valid_session_id(value: &str) -> Option<&str> {
    let value = value.trim();
    if value.is_empty() || value.len() > MAX_SESSION_ID_LEN || value.chars().any(char::is_control) {
        return None;
    }
    Some(value)
}

#[cfg(test)]
mod tests;
